types/next-auth/adapters.d.ts

```diff
@@ -2,50 +2,82 @@ import { ConnectionOptions, EntitySchema } from 'typeorm';
 import { AppOptions } from '.';
 import { SessionProvider } from './client';
 
-// NOTE: There are a lot of `any`. That's because there could be any schema for each entity
-interface Adapter {
-    getAdapter(
-        appOptions: AppOptions,
-    ): Promise<{
-        createUser(profile: any): Promise<any>;
-        getUser(id: string): Promise<any>;
-        getUserByEmail(email: string): Promise<any>;
-        getUserByProviderAccountId(providerId: string, providerAccountId: string): Promise<any>;
-        updateUser(profile: any): Promise<any>;
-        linkAccount(
-            userId: string,
-            providerId: string,
-            providerType: string,
-            providerAccountId: string,
-            refreshToken: string,
-            accessToken: string,
-            accessTokenExpires: number,
-        ): Promise<void>;
-        createSession(user: any): Promise<any>;
-        getSession(sessionToken: string): Promise<any>;
-        updateSession(session: any): Promise<void>;
-        deleteSession(sessionToken: string): Promise<void>;
-        createVerificationRequest?(
-            email: string,
-            url: string,
-            token: string,
-            secret: string,
-            provider: SessionProvider,
-            options: AppOptions,
-        ): Promise<void>;
-        getVerificationRequest?(
-            email: string,
-            verificationToken: string,
-            secret: string,
-            provider: SessionProvider,
-        ): Promise<any>;
-        deleteVerificationRequest?(
-            email: string,
-            verificationToken: string,
-            secret: string,
-            provider: SessionProvider,
-        ): Promise<any>;
-    }>;
+export interface Profile {
+    id: string;
+    name: string;
+    email: string;
+    image: string;
+}
+
+export interface Session {
+    userId: string;
+    expires: Date;
+    sessionToken: string;
+    accessToken: string;
+}
+
+export interface VerificationRequest {
+    identifier: string;
+    token: string;
+    expires: Date;
+}
+
+export interface SendVerificationRequestParams {
+    identifier: string;
+    url: string;
+    token: string;
+    baseUrl: string;
+    provider: SessionProvider;
+}
+
+export type EmailSessionProvider = SessionProvider & {
+    sendVerificationRequest: (params: SendVerificationRequestParams) => Promise<void>;
+    maxAge: number | undefined;
+};
+
+export interface AdapterInstance<TUser, TProfile, TSession, TVerificationRequest> {
```

Comment:
I love it! I think you could extend each generic type: `TProfile` --> `TProfile extends Profile` ([source](https://next-auth.js.org/schemas/adapters#custom-models))

Reply:
Sounds good! I left `TUser` unconstrained since it seemed like that object shape might be a bit more fluid. If you would prefer the constraint there, too, let me know and I can add it.

Reply:
Hmm... maybe `TUser` should extend [this](https://github.com/nextauthjs/next-auth/blob/main/src/adapters/typeorm/models/user.js) and `TProfile` should be `any` by default.

My reasoning is that `TUser` may be used by next-auth in some way, where `TProfile` is only used to create a user. You can create your user using the data you want, but the user created _must_ extend `User`... what do you think?

Reply:
Looking through all of the providers, I think the following should be a good base for `Profile`:
```ts
export interface Profile {
    id: string;
    name: string;
    email: string | null;
    image?: string | null;
}
```

There's a `User` interface already defined that I could have the generic type extend: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/next-auth/index.d.ts#L96-L100

That sound good to you?

Reply:
But then I wonder if the `User` interface should be adjusted to fit the TypeORM definition:
```
interface User {
    name?: string | null;
    email?: string | null;
    image?: string | null;
}
```

Reply:
I think it should, but I'm not sure... @lluia, what do you think?

Reply:
See [this](https://github.com/nextauthjs/next-auth/blob/8115a7c66cdb1d86b0d2a0d76b9aa33c2bfaa33b/src/server/routes/session.js#L67). It uses the TypeORM interface by default

Reply:
I made some updates that I think cover things. If I misunderstood anything or if something needs to be changed, please let me know!

Reply:
Ups sorry saw this late 😓 , looks awesome @jgeurts , thx a lot 💯 💚